lib/std/x/os/Socket.zig

```diff
@@ -23,19 +15,24 @@ pub fn init(domain: u32, socket_type: u32, protocol: u32) !Socket {
     return Socket{ .fd = try os.socket(domain, socket_type, protocol) };
 }
 
+/// Enclose a socket abstraction over an existing socket file descriptor.
+pub fn from(fd: os.socket_t) Socket {
+    return Socket{ .fd = fd };
+}
+
 /// Closes the socket.
 pub fn deinit(self: Socket) void {
     os.closeSocket(self.fd);
 }
 
-/// Shutdown either the read side, or write side, or the entirety of a socket.
+/// Shutdown either the read side, write side, or all side of the socket.
 pub fn shutdown(self: Socket, how: os.ShutdownHow) !void {
     return os.shutdown(self.fd, how);
 }
 
 /// Binds the socket to an address.
-pub fn bind(self: Socket, address: net.Address) !void {
-    return os.bind(self.fd, &address.any, address.getOsSockLen());
+pub fn bind(self: Socket, comptime Address: type, address: Address) !void {
+    return os.bind(self.fd, @ptrCast(*const os.sockaddr, &address), @sizeOf(Address));
```

Comment:
This seems rather unsafe as the API allows passing literally any type which will then be cast to a sockaddr and passed directly to the kernel with no checks. Perhaps it would be better to define a union over the address types and accept that here?

Reply:
Thought about that, though it may run into the issue that @Aransentin brought up in #8590 where the union alone would unnecessarily be 100+ bytes like `std.net.Address`.

My thoughts were, as an alternative, allowing for an arbitrary `Address` type specified would allow for applications that i.e. operate only over TCP/IP to have addresses that only take up 28 bytes. Whereas, applications that operate only over Unix sockets would operate with addresses that take up ~100 bytes.

That being said, I agree that it is somewhat unsafe. Might you have any other alternatives in mind?

Reply:
I'd create a proper `sockaddr` locally from a zig style `Address` type instead of trying to retrofit the zig type into the syscall types. this will not work for all operating systems, as those might not even expose `sockaddr` as a type.

Introduce a function `Adress.toNative_sockaddr()` and create a local temporary `sockaddr`

Reply:
@MasterQ32 Thought about that design, though say for example we have this struct:

```zig
pub const Client = struct {
  address: std.x.net.Address,
  socket: std.x.os.Socket,
  frame: @Frame(Client.readLoop),
};
```

Suppose `Address.toNative_sockaddr()` gets called in `Client.readLoop()`.

This would mean an additional 16 (sockaddr_in) to 28 (sockaddr_in6) bytes would need to be allocated inside `Client` inside of `@Frame(Client.readLoop)` alongside the 28 bytes for the `std.x.net.Address` field inside `Client`.

As an alternative, in the case of `sockaddr` not being exposed as a type in some platforms, some comptime checks can be included to have functions that use `sockaddr` return an error.

In the end, it's overall a tradeoff between the amount of memory we want users to keep around for storing addresses vs. making a safer API. I'm happy to go with any of the approaches though.

Reply:
Went over with @MasterQ32 on his suggested approach. Previous comment I made about additional bytes being redundantly allocated with his approach alongside a `Client` is wrong.

As soon as CI checks out, I'll go with his approach :smile:.

Reply:
Finished implementing the new approach.